source/fe/fe_tools_interpolate.cc

```diff
@@ -712,6 +719,1426 @@ namespace FETools
+                                const InVector                 &u)
+    {
+      std::vector<CellData>  cells_we_need,
+          cells_to_compute,
+          received_cells,
+          received_needs,
+          new_needs,
+          computed_cells,
+          cells_to_send;
+
+      // Compute all the cells needed
+      // from other processes.
+      compute_needs (dof2, cells_we_need);
+
+      // Send the cells needed to there
+      // owners and receive a list of cells other
+      // processes need from us.
+      send_cells (cells_we_need, received_needs);
+
+      // The list of received needs can contain
+      // some cells more than once because different
+      // processes may need data from the same cell.
+      // To compute data only once, generate a vector
+      // with unique entries and distribute the computed
+      // data afterwards back to a vector with correct
+      // receivers.
+      // Computing cell_data can cause a need for
+      // data from some new cells.
+      // If a cell is computed, send it back to
+      // their senders, maybe receive new needs and
+      // compute again, do not wait that all cells
+      // are computed or all needs are collected.
+      // Otherwise we can run into a deadlock if
+      // a cell needed from another process,
+      // itself needs some data from us.
+      unsigned int ready = 0;
+      do
+        {
+          for (unsigned int i=0; i<received_needs.size(); ++i)
+            cell_data_insert (received_needs[i], cells_to_compute);
+
+          compute_cells (dof2, u, cells_to_compute, computed_cells, new_needs);
+
+          // if there are no cells to compute and no new needs, stop
+          ready = Utilities::MPI::sum (new_needs.size () + cells_to_compute.size (), communicator);
+
+          for (typename std::vector<CellData>::const_iterator comp=computed_cells.begin ();
+               comp != computed_cells.end ();
+               ++comp)
+            {
+              // store computed cells
+              cell_data_insert (*comp, available_cells);
+
+              // and generate a vector
+              // of computed cells with
+              // correct receivers
+              // then delete this received
+              // need from the list
+              typename std::vector<CellData>::iterator recv=received_needs.begin();
+              while (recv != received_needs.end())
+                {
+                  if (dealii::internal::p4est::quadrant_is_equal<dim>(recv->quadrant, comp->quadrant))
+                    {
+                      recv->dof_values = comp->dof_values;
+                      cells_to_send.push_back (*recv);
+                      received_needs.erase (recv);
+                      recv = received_needs.begin();
+                    }
+                  else
+                    ++recv;
+                }
+            }
+
+          send_cells (cells_to_send, received_cells);
```

Comment:
You could do this without a barrier if you used tags that change from one iteration to the next, and only accept messages that have the right tag. This way you block only on that one processor, instead of requiring the global barrier.